From 42d601f6839884c534f5089b231c3879944ca7b2 Mon Sep 17 00:00:00 2001
From: Fam Zheng <famz@redhat.com>
Date: Mon, 4 Aug 2014 09:48:48 -0400
Subject: [virt] hyperv/storvsc: Set srb_flags in all cases

Message-id: <1407145729-25017-7-git-send-email-famz@redhat.com>
Patchwork-id: 86957
O-Subject: [RHEL-7 PATCH 6/7] drivers: scsi: storvsc: Set srb_flags in all cases
Bugzilla: 1122317
RH-Acked-by: Jason Wang <jasowang@redhat.com>
RH-Acked-by: Paolo Bonzini <pbonzini@redhat.com>

From: K. Y. Srinivasan <kys@microsoft.com>

Correctly set SRB flags for all valid I/O directions. Some IHV drivers on the
Windows host require this. The host validates the command and SRB flags
prior to passing the command down to native driver stack.

Signed-off-by: K. Y. Srinivasan <kys@microsoft.com>
Reviewed-by: Hannes Reinecke <hare@suse.de>
Cc: <stable@vger.kernel.org>
Signed-off-by: Christoph Hellwig <hch@lst.de>
(cherry picked from commit f885fb73f64154690c2158e813de56363389ffec)
Signed-off-by: Fam Zheng <famz@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/storvsc_drv.c b/drivers/scsi/storvsc_drv.c
index 0a1a31c..d2c7cc8 100644
--- a/drivers/scsi/storvsc_drv.c
+++ b/drivers/scsi/storvsc_drv.c
@@ -1598,26 +1598,24 @@ static int storvsc_queuecommand(struct Scsi_Host *host, struct scsi_cmnd *scmnd)
  vm_srb = &cmd_request->vstor_packet.vm_srb;
  vm_srb->win8_extension.time_out_value = 60;
 
+ vm_srb->win8_extension.srb_flags |=
+  (SRB_FLAGS_QUEUE_ACTION_ENABLE |
+  SRB_FLAGS_DISABLE_SYNCH_TRANSFER);
 
  /* Build the SRB */
  switch (scmnd->sc_data_direction) {
  case DMA_TO_DEVICE:
   vm_srb->data_in = WRITE_TYPE;
   vm_srb->win8_extension.srb_flags |= SRB_FLAGS_DATA_OUT;
-  vm_srb->win8_extension.srb_flags |=
-   (SRB_FLAGS_QUEUE_ACTION_ENABLE |
-   SRB_FLAGS_DISABLE_SYNCH_TRANSFER);
   break;
  case DMA_FROM_DEVICE:
   vm_srb->data_in = READ_TYPE;
   vm_srb->win8_extension.srb_flags |= SRB_FLAGS_DATA_IN;
-  vm_srb->win8_extension.srb_flags |=
-   (SRB_FLAGS_QUEUE_ACTION_ENABLE |
-   SRB_FLAGS_DISABLE_SYNCH_TRANSFER);
   break;
  default:
   vm_srb->data_in = UNKNOWN_TYPE;
-  vm_srb->win8_extension.srb_flags = 0;
+  vm_srb->win8_extension.srb_flags |= (SRB_FLAGS_DATA_IN |
+           SRB_FLAGS_DATA_OUT);
   break;
  }
 
-- 
1.7.1